src/Service/FormDataModelFactory.php

```diff
@@ -0,0 +1,181 @@
+    public function control(
+        Control $control,
+        array $schema,
+        string $name,
+        mixed $value,
+    ): void {
+        if (empty($value)) {
+            if (!$this->includeEmptyFields) {
+                return;
+            }
+            $value = '';
+        }
+
+        $type = $this->identifyType($control, $schema);
+        if ($type === 'file' && !empty($value)) {
+            /** @var string $value */
+            $uploadFile = $this->dataUrlParser->parse($value);
+            /** @var EmailMessageModelFileUpload $value */
+            $value = get_object_vars($uploadFile);
+        }
+
+        $options = null;
+
+        $optionsFromSchema = $schema['items']['oneOf'] ?? $schema['oneOf'] ?? null;
+        if ($optionsFromSchema !== null) {
+            $selected = is_array($value) ? $value : [$value];
+            $value = [];
+            $options = [];
+            foreach ($optionsFromSchema as $option) {
+                $isSelected = in_array($option['const'], $selected, true);
+                $label = $option['title'];
+                $options[] = [
+                    'label' => $label,
+                    'value' => $option['const'],
+                    'selected' => $isSelected,
+                ];
+                if ($isSelected) {
+                    $value[] = $label;
+                }
+            }
+            if (($schema['type'] ?? '') === 'string') {
+                $value = $value[0] ?? '';
+            }
+        }
+
+        $item = [
+            'type' => $type,
+            'name' => $name,
+        ];
+        if (!empty($control->label)) {
+            $item['label'] = $control->label;
+        }
+        if (!empty($control->htmlLabel)) {
+            $item['htmlLabel'] = $control->htmlLabel;
+        }
+        if (!empty($value)) {
+            $item['value'] = $value;
+        }
+        if (!empty($options)) {
+            $item['options'] = $options;
+        }
+
+        $this->items[] = $item;
+    }
+
+    /**
+     * @param JsonSchema $schema
+     */
+    private function identifyType(Control $control, array $schema): string
+    {
+
+        $type = $schema['type'] ?? '';
+        $format = $schema['format'] ?? '';
+
+        if ($type === 'string' && $format === 'data-url') {
+            return 'file';
+        }
+
+        if ($type === 'string' && $format === 'html') {
+            return 'html';
+        }
+
+        if ($type === 'boolean') {
+            return 'checkbox';
+        }
+
+        if ($type === 'array' && isset($schema['items']['oneOf'])) {
+            return 'checkbox-group';
+        }
+
+        if (
+            $type === 'string'
+            && isset($schema['oneOf'])
+            && ($control->options['format'] ?? '') === 'radio') {
+            return 'radio-buttons';
+        }
+
+        if ($type === 'string' && isset($schema['oneOf'])
+        ) {
+            return 'select';
+        }
+
+        return 'text';
+    }
```

Comment:
Yes, text is the fallback.

UISchema and Json schema are used to identify which UI component is to be used.

See https://sitepark.github.io/atoolo-docs/develop/form/controls/#json-schema

Without fallback, it would mean that new components can only be generated once the FEDS module has been updated. With fallback, it makes the system more stable during an update process.